### PR TITLE
[Feature] support map functions map_size(), map_keys(), map_values()

### LINKED ISSUE #10677: Support struct/map for Data Lake
This issue is to lay the groundwork for supporting complex types (struct/map) for StarRocks on data lakes.

The goals of this issue:
- support complex types (struct/map) for hive/hudi/iceberg
- support reading complex types (struct/map) from orc/parquet/csv
- support map functions
- support pushdown subfields of struct

The subtasks for this issue:
- [ ] support reading full map column from hive/orc https://github.com/StarRocks/starrocks/pull/10546
- [ ] support reading full struct column from hive/orc
- [ ] support `[]` operator for map type
- [ ] support `.` operator for struct type
- [ ] support pushdown subfields of struct type
- [ ] support map functions
- [ ] support reading map/struct column from parquet/csv
- [ ] support reading map/struct column from iceberg/hudi

### PR DESCRIPTION
**Pull request by zombee0**
Signed-off-by: zombee0 <flylucas_10@163.com>

## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #10677

## Problem Summary(Required) ：

1. In fe I defined a new type AnyMapType which is used to support the PolymorphicFunction.
2. In Analyze phase, It will check the type of argument,  and set the real type of return type.

the use case may like these:
`-- map_size()
select map_size(c2) from map_tbl;
select c1 from map_tbl where map_size(c2) > 2;

-- map_keys()
select map_keys(c2) from map_tbl;
select c1 from map_tbl where array_contains(map_keys(c2), "k1");

-- map_values()
select map_values(c2) from map_tbl;
select c1 from map_tbl where array_contains(map_values(c2), "v1");`

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 39 (38.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MapType.java|0|4|00.00%|[64, 65, 67, 68]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|10|29|34.48%|[1146, 1147, 1149, 1150, 1151, 1152, 1153, 1195, 1196, 1197, 1198, 1199, 1201, 1202, 1209, 1219, 1290, 1291, 1292]|
|:large_blue_circle:|com/starrocks/catalog/AnyMapType.java|2|3|66.67%|[13]|
|:large_blue_circle:|com/starrocks/catalog/PseudoType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 39 (38.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MapType.java|0|4|00.00%|[64, 65, 67, 68]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|10|29|34.48%|[1146, 1147, 1149, 1150, 1151, 1152, 1153, 1195, 1196, 1197, 1198, 1199, 1201, 1202, 1209, 1219, 1290, 1291, 1292]|
|:large_blue_circle:|com/starrocks/catalog/AnyMapType.java|2|3|66.67%|[13]|
|:large_blue_circle:|com/starrocks/catalog/PseudoType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 39 (38.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MapType.java|0|4|00.00%|[64, 65, 67, 68]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|10|29|34.48%|[1146, 1147, 1149, 1150, 1151, 1152, 1153, 1195, 1196, 1197, 1198, 1199, 1201, 1202, 1209, 1219, 1290, 1291, 1292]|
|:large_blue_circle:|com/starrocks/catalog/AnyMapType.java|2|3|66.67%|[13]|
|:large_blue_circle:|com/starrocks/catalog/PseudoType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 39 (38.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MapType.java|0|4|00.00%|[64, 65, 67, 68]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|10|29|34.48%|[1161, 1162, 1164, 1165, 1166, 1167, 1168, 1210, 1211, 1212, 1213, 1214, 1216, 1217, 1224, 1234, 1305, 1306, 1307]|
|:large_blue_circle:|com/starrocks/catalog/AnyMapType.java|2|3|66.67%|[13]|
|:large_blue_circle:|com/starrocks/catalog/PseudoType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by wangruin** [comment]
run starrocks_be_unittest

**Comment by zombee0** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 39 (38.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MapType.java|0|4|00.00%|[64, 65, 67, 68]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|10|29|34.48%|[1161, 1162, 1164, 1165, 1166, 1167, 1168, 1210, 1211, 1212, 1213, 1214, 1216, 1217, 1224, 1234, 1305, 1306, 1307]|
|:large_blue_circle:|com/starrocks/catalog/AnyMapType.java|2|3|66.67%|[13]|
|:large_blue_circle:|com/starrocks/catalog/PseudoType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 26 / 39 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MapType.java|2|4|50.00%|[67, 68]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|19|29|65.52%|[1162, 1166, 1167, 1168, 1216, 1217, 1224, 1305, 1306, 1307]|
|:large_blue_circle:|com/starrocks/catalog/AnyMapType.java|2|3|66.67%|[13]|
|:large_blue_circle:|com/starrocks/catalog/PseudoType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Review by DorianZheng — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java [hunk 1]
```diff
@@ -1156,6 +1157,16 @@ private Function checkPolymorphicFunction(Function fn, Type[] paramTypes) {
                     LOGGER.warn("could not determine polymorphic type because input has non-match types");
                     return null;
                 }
+            } else if (declType instanceof AnyMapType) {
+                if (realType.isNull()) {
+                    continue;
+                }
+                if (typeMap == null) {
+                    typeMap = (MapType) realType;
+                } else if ((typeMap = (MapType) getSuperType(typeMap, realType)) == null) {
```
Comment (Smith-Cruise) [inline fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java:1166]: I think this step is used to check the function all parameters are the same type(including nested types in complex type), so our function must apply the totally same type for complex type?

**Review by Youngwb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java (hunk 1 above)
Comment (zombee0) [reply fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java:1166]: up to now, it's impossible, i will rewrite it.

File: be/src/exprs/vectorized/map_functions.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+ColumnPtr MapFunctions::map_keys(starrocks_udf::FunctionContext* context, const Columns& columns) {
+    DCHECK_EQ(1, columns.size());
+    RETURN_IF_COLUMNS_ONLY_NULL(columns);
+
+    Column* arg0 = columns[0].get();
+
+    auto* col_map = down_cast<MapColumn*>(ColumnHelper::get_data_column(arg0));
+    auto map_keys = col_map->keys().clone();
+    auto map_keys_array = ArrayColumn::create(std::move(map_keys), UInt32Column::create(col_map->offsets()));
```
[the remark this answers is not in the document]
Comment (zombee0) [reply be/src/exprs/vectorized/map_functions.cpp:45]: got it

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 38 (39.47%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MapType.java|0|4|00.00%|[64, 65, 67, 68]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|10|28|35.71%|[1174, 1175, 1177, 1178, 1180, 1181, 1223, 1224, 1225, 1226, 1227, 1229, 1230, 1237, 1247, 1318, 1319, 1320]|
|:large_blue_circle:|com/starrocks/catalog/AnyMapType.java|2|3|66.67%|[13]|
|:large_blue_circle:|com/starrocks/catalog/PseudoType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by zombee0** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 26 / 38 (68.42%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MapType.java|2|4|50.00%|[67, 68]|
|:large_blue_circle:|com/starrocks/catalog/AnyMapType.java|2|3|66.67%|[13]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|19|28|67.86%|[1175, 1180, 1181, 1229, 1230, 1237, 1318, 1319, 1320]|
|:large_blue_circle:|com/starrocks/catalog/PseudoType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Review by Youngwb — APPROVED** [review]

**Review by Smith-Cruise — APPROVED** [review]
LGTM

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**